dawn/src/dawn/CodeGen/Cuda-ico/CudaIcoCodeGen.cpp

```diff
@@ -908,6 +871,61 @@ void CudaIcoCodeGen::generateAllAPIRunFunctions(
   }
 }
 
+void CudaIcoCodeGen::generateMemMgmtFunctions(
+    std::stringstream& ssSW, const std::shared_ptr<iir::StencilInstantiation>& stencilInstantiation,
+    CodeGenProperties& codeGenProperties, bool onlyDecl) const {
+  const std::string wrapperName = stencilInstantiation->getName();
+  std::string stencilName = codeGenProperties.getStencilName(
+      StencilContext::SC_Stencil, stencilInstantiation->getStencils()[0]->getStencilID());
+  const std::string fullStencilName =
+      "dawn_generated::cuda_ico::" + wrapperName + "::" + stencilName;
+
+  MemberFunction setupFun("void", "setup_" + wrapperName, ssSW, 0, onlyDecl);
+  setupFun.addArg("dawn::GlobalGpuTriMesh *mesh");
+  setupFun.addArg("int k_size");
+  setupFun.finishArgs();
+  if(!onlyDecl) {
+    setupFun.addStatement(fullStencilName + "::setup(mesh, k_size)");
+  }
+  setupFun.commit();
+  if(onlyDecl) {
+    ssSW << ";";
+  }
+
+  MemberFunction freeFun("void", "free_" + wrapperName, ssSW, 0, onlyDecl);
+  freeFun.finishArgs();
+  if(!onlyDecl) {
+    freeFun.startBody();
+    freeFun.addStatement(fullStencilName + "::free()");
+  }
+  freeFun.commit();
+  if(onlyDecl) {
+    ssSW << ";";
```

Comment:
A bit out of scope of the PR but it could be solved here: the generated C header has semicolons on new lines, which still works but it's pretty weird.